tests/kernel/mem_protect/mem_protect/src/mem_domain.c

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "mem_protect.h"
+
+#define ERROR_STR \
+	("Fault didn't occur when we accessed a unassigned memory domain.")
+
+/****************************************************************************/
+K_THREAD_STACK_DEFINE(mem_domain_1_stack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(mem_domain_2_stack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(mem_domain_6_stack, MEM_DOMAIN_STACK_SIZE);
+__kernel struct k_thread mem_domain_1_tid, mem_domain_2_tid, mem_domain_6_tid;
+
+/****************************************************************************/
+/* The mem domains needed.*/
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_buf[MEM_REGION_ALLOC];
+
+/* partitions added later in the test cases.*/
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part1[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part2[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part3[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part4[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part5[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part6[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part7[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part8[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(mem_domain_memory_partition,
+		       mem_domain_buf,
+		       sizeof(mem_domain_buf),
+		       K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *mem_domain_memory_partition_array[] = {
+	&mem_domain_memory_partition
+};
+
+struct k_mem_domain mem_domain_mem_domain;
+
+/****************************************************************************/
+/* Common init functions */
+static inline void mem_domain_init(void)
+{
+	k_mem_domain_init(&mem_domain_mem_domain,
+			  MEM_PARTITION_INIT_NUM,
+			  mem_domain_memory_partition_array);
+}
+
+static inline void set_valid_fault_value(int test_case_number)
+{
+	switch (test_case_number) {
+	case 1:
+		valid_fault = false;
+		break;
+	case 2:
+		valid_fault = true;
+		break;
+	default:
+		valid_fault = false;
+		break;
+	}
+}
+
+/****************************************************************************/
+/* Userspace function */
+void mem_domain_for_user(void *tc_number, void *p2, void *p3)
+{
+	set_valid_fault_value((u32_t)tc_number);
+
+	mem_domain_buf[0] = 10;
```

Comment:
Don't you need to disable CONFIG_APPLICATION_MEMORY for this test to work?
Otherwise, mem_domain_buf is part of application global memory and accessible to all usermode threads by default.  How does this pass on qemu_x86?

Reply:
It passes because since memory domains can only be configured in supervisor mode, there is no system call interface for them and hence they are not treated as kernel objects that are validated.

I am contemplating what to do about this since this seems an easy programming mistake. We could at least put some assertions in the code that the struct isn't in a user writable region.

CONFIG_APPLICATION_MEMORY was an easy and convenient way to get test cases working, but I am less convinced that it is suitable for real-world development as time goes by.
I feel we need something better, something that ties into the memory domain APIs and lets you arrange data in a logical fashion rather than having to do stuff by hand in the linker scripts, and also support the notion of multiple logical "applications" which happen to share the same MCU.

Reply:
This works because once the memory domains partitions are configured that memory region will be accessible to only the threads in the memory domain. It doesn't matter it if is in _app memory or in _kernel memory.

And I agree I should have added  __kernel macro to it. This would have made it more sensible.

And Andrew maybe I can add one patch to make the linker always force the memory domain structures to kernel data region. This should protect the structs which are in app memory.
In future when we change the CONFIG_APPLICATION_MEMORY we can modify this.